File: 14_lmm_sabr_hybrid_model/python/lmm_sabr_hybrid.py
"""
lmm_sabr_hybrid.py

LMM-SABR hybrid model under the terminal measure.

The model reuses the Chapter 13 components:

    TenorStructure
    GaussianLMMVolatility
    correlation loadings
    PCA factor reduction
    terminal-measure drift structure
    discount-factor reconstruction

The forward-rate dynamics are:

    dL_i(t)
        =
        mu_i(t) dt
        +
        lambda_i(t) . dW_t

where:

    lambda_i(t)
        =
        g_i(t)
        alpha_i(t)
        [L_i(t) + shift_i]^beta_i
        b_i

The stochastic-volatility dynamics are:

    d alpha_i(t)
        =
        nu_i alpha_i(t) dZ_i(t)

with correlation:

    Corr(dW_i, dZ_i) = rho_i.
"""

import numpy as np


class LMMSABRHybrid:
    """
    Terminal-measure LMM-SABR hybrid model.
    """

    def __init__(
        self,
        tenor_structure,
        initial_forwards,
        deterministic_volatility_model,
        correlation_loadings,
        initial_discount_factors,
        sabr_parameters
    ):

        self.tenor = tenor_structure

        self.initial_forwards = np.asarray(
            initial_forwards,
            dtype=float
        )

        self.deterministic_volatility_model = (
            deterministic_volatility_model
        )

        self.correlation_loadings = np.asarray(
            correlation_loadings,
            dtype=float
        )

        self.initial_discount_factors = np.asarray(
            initial_discount_factors,
            dtype=float
        )

        self.sabr_parameters = (
            sabr_parameters
        )

        self.number_of_forwards = len(
            self.initial_forwards
        )

        if self.correlation_loadings.ndim != 2:

            raise ValueError(
                "correlation_loadings must be a matrix."
            )

        self.number_of_factors = (
            self.correlation_loadings.shape[1]
        )

        self.validate_inputs()

    def validate_inputs(
        self
    ):
        """
        Validate model dimensions and initial values.
        """

        if self.number_of_forwards <= 0:

            raise ValueError(
                "At least one forward rate is required."
            )

        if (
            self.correlation_loadings.shape[0]
            !=
            self.number_of_forwards
        ):

            raise ValueError(
                "Correlation loading rows must equal "
                "the number of forwards."
            )

        if self.number_of_factors <= 0:

            raise ValueError(
                "At least one rate factor is required."
            )

        if (
            len(
                self.initial_discount_factors
            )
            !=
            self.number_of_forwards + 1
        ):

            raise ValueError(
                "Initial discount factors must have length "
                "number_of_forwards + 1."
            )

        if np.any(
            self.initial_discount_factors <= 0.0
        ):

            raise ValueError(
                "Initial discount factors must be positive."
            )

        if (
            self.sabr_parameters.number_of_forwards
            !=
            self.number_of_forwards
        ):

            raise ValueError(
                "SABR parameters are inconsistent "
                "with the forward vector."
            )

        accruals = self.get_accruals()

        if len(accruals) != self.number_of_forwards:

            raise ValueError(
                "Tenor accruals are inconsistent "
                "with the forward vector."
            )

        if np.any(
            1.0
            +
            accruals
            *
            self.initial_forwards
            <=
            0.0
        ):

            raise ValueError(
                "Initial forwards violate "
                "1 + delta * L > 0."
            )

        shifted_forwards = (
            self.initial_forwards
            +
            self.sabr_parameters.shift
        )

        if np.any(
            shifted_forwards <= 0.0
        ):

            raise ValueError(
                "Initial forwards plus shifts "
                "must be positive."
            )

    def get_reset_times(
        self
    ):
        """
        Return forward reset times.
        """

        return np.asarray(
            self.tenor.times[:-1],
            dtype=float
        )

    def get_accruals(
        self
    ):
        """
        Return accrual periods.
        """

        return np.asarray(
            self.tenor.accruals,
            dtype=float
        )

    def active_mask(
        self,
        time
    ):
        """
        Identify forwards that have not yet reset.
        """

        return (
            float(time)
            <
            self.get_reset_times()
            -
            1.0e-12
        )

    def deterministic_volatilities(
        self,
        time
    ):
        """
        Reuse the Chapter 13 deterministic volatility term structure.
        """

        return np.asarray(
            self.deterministic_volatility_model
            .instantaneous_vols(
                time
            ),
            dtype=float
        )

    def validate_state(
        self,
        forwards,
        alpha
    ):
        """
        Validate a simulated model state.
        """

        forwards = np.asarray(
            forwards,
            dtype=float
        )

        alpha = np.asarray(
            alpha,
            dtype=float
        )

        if forwards.shape != (
            self.number_of_forwards,
        ):

            raise ValueError(
                "Forward state has the wrong shape."
            )

        if alpha.shape != (
            self.number_of_forwards,
        ):

            raise ValueError(
                "Alpha state has the wrong shape."
            )

        if not np.all(
            np.isfinite(
                forwards
            )
        ):

            raise RuntimeError(
                "Forward state contains invalid values."
            )

        if not np.all(
            np.isfinite(
                alpha
            )
        ):

            raise RuntimeError(
                "Alpha state contains invalid values."
            )

        if np.any(
            alpha <= 0.0
        ):

            raise RuntimeError(
                "All alpha values must be positive."
            )

        if np.any(
            1.0
            +
            self.get_accruals()
            *
            forwards
            <=
            0.0
        ):

            raise RuntimeError(
                "Forward state violates "
                "1 + delta * L > 0."
            )

        if np.any(
            forwards
            +
            self.sabr_parameters.shift
            <=
            0.0
        ):

            raise RuntimeError(
                "Forward state violates "
                "L + shift > 0."
            )

    def factor_loadings(
        self,
        time,
        forwards,
        alpha
    ):
        """
        Calculate SABR-adjusted forward factor loadings.

        Returns
        -------
        numpy.ndarray
            Matrix with shape:

            number_of_forwards
            x
            number_of_factors
        """

        forwards = np.asarray(
            forwards,
            dtype=float
        )

        alpha = np.asarray(
            alpha,
            dtype=float
        )

        self.validate_state(
            forwards,
            alpha
        )

        shifted_forwards = (
            forwards
            +
            self.sabr_parameters.shift
        )

        cev_term = np.power(
            shifted_forwards,
            self.sabr_parameters.beta
        )

        volatility_scale = (
            self.deterministic_volatilities(
                time
            )
            *
            alpha
            *
            cev_term
        )

        active = self.active_mask(
            time
        )

        volatility_scale = np.where(
            active,
            volatility_scale,
            0.0
        )

        return (
            volatility_scale[:, None]
            *
            self.correlation_loadings
        )

    def instantaneous_covariance_matrix(
        self,
        time,
        forwards,
        alpha
    ):
        """
        Calculate the instantaneous forward covariance matrix.
        """

        loadings = self.factor_loadings(
            time,
            forwards,
            alpha
        )

        return (
            loadings
            @
            loadings.T
        )

    def terminal_measure_drift(
        self,
        time,
        forwards,
        alpha
    ):
        """
        Calculate terminal-measure forward drifts.

        The drift is:

            mu_i
                =
                - sum over j > i
                  delta_j
                  lambda_i . lambda_j
                  /
                  (1 + delta_j L_j)
        """

        forwards = np.asarray(
            forwards,
            dtype=float
        )

        alpha = np.asarray(
            alpha,
            dtype=float
        )

        self.validate_state(
            forwards,
            alpha
        )

        loadings = self.factor_loadings(
            time,
            forwards,
            alpha
        )

        active = self.active_mask(
            time
        )

        accruals = self.get_accruals()

        drift = np.zeros(
            self.number_of_forwards
        )

        for i in range(
            self.number_of_forwards
        ):

            if not active[i]:

                continue

            total = 0.0

            for j in range(
                i + 1,
                self.number_of_forwards
            ):

                if not active[j]:

                    continue

                denominator = (
                    1.0
                    +
                    accruals[j]
                    *
                    forwards[j]
                )

                if denominator <= 0.0:

                    raise RuntimeError(
                        "Invalid denominator in drift calculation."
                    )

                covariance = np.dot(
                    loadings[i],
                    loadings[j]
                )

                total += (
                    accruals[j]
                    *
                    covariance
                    /
                    denominator
                )

            drift[i] = -total

        return drift

    def effective_forward_shocks(
        self,
        time,
        forwards,
        alpha,
        rate_factor_shocks
    ):
        """
        Convert common factor shocks into one normalized shock
        for every forward rate.

        These shocks are used to create correlation between:

            forward-rate changes

        and:

            alpha changes.
        """

        rate_factor_shocks = np.asarray(
            rate_factor_shocks,
            dtype=float
        )

        if rate_factor_shocks.shape != (
            self.number_of_factors,
        ):

            raise ValueError(
                "rate_factor_shocks has the wrong shape."
            )

        loadings = self.factor_loadings(
            time,
            forwards,
            alpha
        )

        raw_forward_shocks = (
            loadings
            @
            rate_factor_shocks
        )

        loading_norms = np.linalg.norm(
            loadings,
            axis=1
        )

        normalized_shocks = np.zeros(
            self.number_of_forwards
        )

        non_zero = (
            loading_norms > 1.0e-16
        )

        normalized_shocks[non_zero] = (
            raw_forward_shocks[non_zero]
            /
            loading_norms[non_zero]
        )

        return normalized_shocks

    def evolve(
        self,
        time,
        forwards,
        alpha,
        dt,
        rate_factor_shocks,
        independent_volatility_shocks
    ):
        """
        Advance forwards and stochastic volatilities by one time step.

        Forward rates use an Euler step.

        Alpha uses an exact lognormal step.
        """

        if dt <= 0.0:

            raise ValueError(
                "dt must be positive."
            )

        forwards = np.asarray(
            forwards,
            dtype=float
        )

        alpha = np.asarray(
            alpha,
            dtype=float
        )

        rate_factor_shocks = np.asarray(
            rate_factor_shocks,
            dtype=float
        )

        independent_volatility_shocks = np.asarray(
            independent_volatility_shocks,
            dtype=float
        )

        self.validate_state(
            forwards,
            alpha
        )

        if rate_factor_shocks.shape != (
            self.number_of_factors,
        ):

            raise ValueError(
                "rate_factor_shocks has the wrong shape."
            )

        if (
            independent_volatility_shocks.shape
            !=
            (
                self.number_of_forwards,
            )
        ):

            raise ValueError(
                "independent_volatility_shocks "
                "has the wrong shape."
            )

        active = self.active_mask(
            time
        )

        loadings = self.factor_loadings(
            time,
            forwards,
            alpha
        )

        drift = self.terminal_measure_drift(
            time,
            forwards,
            alpha
        )

        square_root_dt = np.sqrt(
            dt
        )

        diffusion = (
            loadings
            @
            rate_factor_shocks
            *
            square_root_dt
        )
        
        next_forwards = (
            forwards
            +
            drift
            *
            dt
            +
            diffusion
        )
        
        next_forwards = np.where(
            active,
            next_forwards,
            forwards
        )
        
        # Keep the shifted forward strictly positive.
        minimum_forwards = (
            -self.sabr_parameters.shift
            +
            1.0e-8
        )
        
        next_forwards = np.where(
            active,
            np.maximum(
                next_forwards,
                minimum_forwards
            ),
            forwards
        )

        effective_rate_shocks = (
            self.effective_forward_shocks(
                time,
                forwards,
                alpha,
                rate_factor_shocks
            )
        )

        rho = self.sabr_parameters.rho

        correlated_volatility_shocks = (
            rho
            *
            effective_rate_shocks
            +
            np.sqrt(
                1.0
                -
                rho
                *
                rho
            )
            *
            independent_volatility_shocks
        )

        nu = self.sabr_parameters.nu

        alpha_log_increment = (
            -0.5
            *
            nu
            *
            nu
            *
            dt
            +
            nu
            *
            square_root_dt
            *
            correlated_volatility_shocks
        )

        next_alpha = (
            alpha
            *
            np.exp(
                alpha_log_increment
            )
        )

        next_alpha = np.clip(
            next_alpha,
            self.sabr_parameters.alpha_floor,
            self.sabr_parameters.alpha_cap
        )

        next_alpha = np.where(
            active,
            next_alpha,
            alpha
        )

        self.validate_state(
            next_forwards,
            next_alpha
        )

        return (
            next_forwards,
            next_alpha
        )

    def discount_factors_from_forwards(
        self,
        forwards,
        start_index
    ):
        """
        Reconstruct discount factors from a forward curve.

        Starting from:

            P(T_k, T_k) = 1

        use:

            P(T_k, T_{j+1})
                =
                P(T_k, T_j)
                /
                (1 + delta_j L_j)
        """

        forwards = np.asarray(
            forwards,
            dtype=float
        )

        if forwards.shape != (
            self.number_of_forwards,
        ):

            raise ValueError(
                "Forward state has the wrong shape."
            )

        if (
            start_index < 0
            or
            start_index > self.number_of_forwards
        ):

            raise ValueError(
                "start_index is outside the tenor grid."
            )

        discount_factors = np.full(
            self.number_of_forwards + 1,
            np.nan
        )

        discount_factors[start_index] = 1.0

        running_discount_factor = 1.0

        accruals = self.get_accruals()

        for j in range(
            start_index,
            self.number_of_forwards
        ):

            denominator = (
                1.0
                +
                accruals[j]
                *
                forwards[j]
            )

            if denominator <= 0.0:

                raise RuntimeError(
                    "Cannot reconstruct discount factors "
                    "from an invalid forward state."
                )

            running_discount_factor /= (
                denominator
            )

            discount_factors[j + 1] = (
                running_discount_factor
            )

        return discount_factors

    def terminal_bond_from_forwards(
        self,
        forwards,
        start_index
    ):
        """
        Return P(T_k, T_N).
        """

        discount_factors = (
            self.discount_factors_from_forwards(
                forwards,
                start_index
            )
        )

        return float(
            discount_factors[-1]
        )

    def initial_terminal_discount_factor(
        self
    ):
        """
        Return P(0, T_N).
        """

        return float(
            self.initial_discount_factors[-1]
        )

    def summary(
        self
    ):
        """
        Return a compact model summary.
        """

        return {
            "model":
                "LMM-SABR Hybrid",

            "number_of_forwards":
                self.number_of_forwards,

            "number_of_rate_factors":
                self.number_of_factors,

            "initial_terminal_discount_factor":
                self.initial_terminal_discount_factor(),

            "sabr_parameters":
                self.sabr_parameters.summary()
        }

File: 14_lmm_sabr_hybrid_model/cpp/LMMSABRHybrid.cpp
#include "LMMSABRHybrid.hpp"

#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits>
#include <stdexcept>


LMMSABRHybrid::LMMSABRHybrid(
    const Vector& resetTimes,
    const Vector& accruals,
    const Vector& initialForwards,
    const Vector& initialDiscountFactors,
    const Matrix& correlationLoadings,
    const SABRParameters& sabrParameters,
    double volatilityLevel,
    double volatilityDecay,
    double volatilityFloor
)
    :
    resetTimes_(resetTimes),
    accruals_(accruals),
    initialForwards_(initialForwards),
    initialDiscountFactors_(
        initialDiscountFactors
    ),
    correlationLoadings_(
        correlationLoadings
    ),
    sabrParameters_(
        sabrParameters
    ),
    volatilityLevel_(
        volatilityLevel
    ),
    volatilityDecay_(
        volatilityDecay
    ),
    volatilityFloor_(
        volatilityFloor
    ),
    numberOfForwards_(
        initialForwards.size()
    ),
    numberOfFactors_(
        correlationLoadings.empty()
        ?
        0
        :
        correlationLoadings.front().size()
    )
{
    validateModelInputs();

    validateState(
        initialForwards_,
        sabrParameters_.alpha0()
    );
}


std::size_t LMMSABRHybrid::numberOfForwards() const
{
    return numberOfForwards_;
}


std::size_t LMMSABRHybrid::numberOfFactors() const
{
    return numberOfFactors_;
}


const Vector& LMMSABRHybrid::resetTimes() const
{
    return resetTimes_;
}


const Vector& LMMSABRHybrid::accruals() const
{
    return accruals_;
}


const Vector& LMMSABRHybrid::initialForwards() const
{
    return initialForwards_;
}


const Vector&
LMMSABRHybrid::initialDiscountFactors() const
{
    return initialDiscountFactors_;
}


const SABRParameters&
LMMSABRHybrid::sabrParameters() const
{
    return sabrParameters_;
}


double LMMSABRHybrid::dotProduct(
    const Vector& first,
    const Vector& second
)
{
    if (first.size() != second.size())
    {
        throw std::invalid_argument(
            "Dot-product vectors have different sizes."
        );
    }

    double result = 0.0;

    for (std::size_t i = 0;
         i < first.size();
         ++i)
    {
        result += first[i] * second[i];
    }

    return result;
}


double LMMSABRHybrid::vectorNorm(
    const Vector& values
)
{
    return std::sqrt(
        dotProduct(
            values,
            values
        )
    );
}


void LMMSABRHybrid::validateModelInputs() const
{
    if (numberOfForwards_ == 0)
    {
        throw std::invalid_argument(
            "The model must contain forward rates."
        );
    }

    if (resetTimes_.size() !=
        numberOfForwards_)
    {
        throw std::invalid_argument(
            "resetTimes size must equal "
            "the number of forwards."
        );
    }

    if (accruals_.size() !=
        numberOfForwards_)
    {
        throw std::invalid_argument(
            "accruals size must equal "
            "the number of forwards."
        );
    }

    if (initialDiscountFactors_.size() !=
        numberOfForwards_ + 1)
    {
        throw std::invalid_argument(
            "initialDiscountFactors must contain "
            "one more value than initialForwards."
        );
    }

    if (correlationLoadings_.size() !=
        numberOfForwards_)
    {
        throw std::invalid_argument(
            "correlationLoadings must contain "
            "one row per forward."
        );
    }

    if (numberOfFactors_ == 0)
    {
        throw std::invalid_argument(
            "At least one rate factor is required."
        );
    }

    for (const Vector& row :
         correlationLoadings_)
    {
        if (row.size() !=
            numberOfFactors_)
        {
            throw std::invalid_argument(
                "All correlation-loading rows "
                "must have equal length."
            );
        }
    }

    if (sabrParameters_.numberOfForwards() !=
        numberOfForwards_)
    {
        throw std::invalid_argument(
            "SABR parameter dimension does not "
            "match the forward dimension."
        );
    }

    if (volatilityLevel_ < 0.0 ||
        volatilityDecay_ < 0.0 ||
        volatilityFloor_ < 0.0)
    {
        throw std::invalid_argument(
            "Volatility parameters must "
            "be non-negative."
        );
    }
}


Vector LMMSABRHybrid::activeMask(
    double time
) const
{
    Vector active(
        numberOfForwards_,
        0.0
    );

    for (std::size_t i = 0;
         i < numberOfForwards_;
         ++i)
    {
        if (time <
            resetTimes_[i] - 1.0e-12)
        {
            active[i] = 1.0;
        }
    }

    return active;
}


Vector LMMSABRHybrid::deterministicVolatilities(
    double time
) const
{
    Vector volatilities(
        numberOfForwards_,
        0.0
    );

    Vector active =
        activeMask(
            time
        );

    for (std::size_t i = 0;
         i < numberOfForwards_;
         ++i)
    {
        if (active[i] > 0.5)
        {
            double timeToReset =
                std::max(
                    resetTimes_[i] - time,
                    0.0
                );

            volatilities[i] =
                volatilityFloor_
                +
                volatilityLevel_
                *
                std::exp(
                    -volatilityDecay_
                    *
                    timeToReset
                );
        }
    }

    return volatilities;
}


Matrix LMMSABRHybrid::factorLoadings(
    double time,
    const Vector& forwards,
    const Vector& alpha
) const
{
    validateState(
        forwards,
        alpha
    );

    Vector deterministicVols =
        deterministicVolatilities(
            time
        );

    Vector active =
        activeMask(
            time
        );

    Matrix loadings(
        numberOfForwards_,
        Vector(
            numberOfFactors_,
            0.0
        )
    );

    for (std::size_t i = 0;
         i < numberOfForwards_;
         ++i)
    {
        if (active[i] < 0.5)
        {
            continue;
        }

        double shiftedForward =
            forwards[i]
            +
            sabrParameters_.shift()[i];

        double cevTerm =
            std::pow(
                shiftedForward,
                sabrParameters_.beta()[i]
            );

        double volatilityScale =
            deterministicVols[i]
            *
            alpha[i]
            *
            cevTerm;

        for (std::size_t factor = 0;
             factor < numberOfFactors_;
             ++factor)
        {
            loadings[i][factor] =
                volatilityScale
                *
                correlationLoadings_[i][factor];
        }
    }

    return loadings;
}


Matrix LMMSABRHybrid::
instantaneousCovarianceMatrix(
    double time,
    const Vector& forwards,
    const Vector& alpha
) const
{
    Matrix loadings =
        factorLoadings(
            time,
            forwards,
            alpha
        );

    Matrix covariance(
        numberOfForwards_,
        Vector(
            numberOfForwards_,
            0.0
        )
    );

    for (std::size_t i = 0;
         i < numberOfForwards_;
         ++i)
    {
        for (std::size_t j = 0;
             j < numberOfForwards_;
             ++j)
        {
            covariance[i][j] =
                dotProduct(
                    loadings[i],
                    loadings[j]
                );
        }
    }

    return covariance;
}


Vector LMMSABRHybrid::terminalMeasureDrift(
    double time,
    const Vector& forwards,
    const Vector& alpha
) const
{
    validateState(
        forwards,
        alpha
    );

    Matrix loadings =
        factorLoadings(
            time,
            forwards,
            alpha
        );

    Vector active =
        activeMask(
            time
        );

    Vector drift(
        numberOfForwards_,
        0.0
    );

    for (std::size_t i = 0;
         i < numberOfForwards_;
         ++i)
    {
        if (active[i] < 0.5)
        {
            continue;
        }

        double total = 0.0;

        for (std::size_t j = i + 1;
             j < numberOfForwards_;
             ++j)
        {
            if (active[j] < 0.5)
            {
                continue;
            }

            double denominator =
                1.0
                +
                accruals_[j]
                *
                forwards[j];

            if (denominator <= 0.0)
            {
                throw std::runtime_error(
                    "Invalid LMM denominator: "
                    "1 + delta * L <= 0."
                );
            }

            double covariance =
                dotProduct(
                    loadings[i],
                    loadings[j]
                );

            total +=
                accruals_[j]
                *
                covariance
                /
                denominator;
        }

        drift[i] = -total;
    }

    return drift;
}


Vector LMMSABRHybrid::effectiveForwardShocks(
    double time,
    const Vector& forwards,
    const Vector& alpha,
    const Vector& rateFactorShocks
) const
{
    if (rateFactorShocks.size() !=
        numberOfFactors_)
    {
        throw std::invalid_argument(
            "Incorrect number of rate-factor shocks."
        );
    }

    Matrix loadings =
        factorLoadings(
            time,
            forwards,
            alpha
        );

    Vector shocks(
        numberOfForwards_,
        0.0
    );

    for (std::size_t i = 0;
         i < numberOfForwards_;
         ++i)
    {
        double norm =
            vectorNorm(
                loadings[i]
            );

        if (norm > 1.0e-14)
        {
            shocks[i] =
                dotProduct(
                    loadings[i],
                    rateFactorShocks
                )
                /
                norm;
        }
    }

    return shocks;
}


void LMMSABRHybrid::evolve(
    double time,
    const Vector& forwards,
    const Vector& alpha,
    double dt,
    const Vector& rateFactorShocks,
    const Vector& independentVolatilityShocks,
    Vector& nextForwards,
    Vector& nextAlpha
) const
{
    if (dt <= 0.0)
    {
        throw std::invalid_argument(
            "The simulation time step must be positive."
        );
    }

    if (rateFactorShocks.size() !=
        numberOfFactors_)
    {
        throw std::invalid_argument(
            "Incorrect rate-factor shock dimension."
        );
    }

    if (independentVolatilityShocks.size() !=
        numberOfForwards_)
    {
        throw std::invalid_argument(
            "Incorrect volatility-shock dimension."
        );
    }

    validateState(
        forwards,
        alpha
    );

    Vector active =
        activeMask(
            time
        );

    Matrix loadings =
        factorLoadings(
            time,
            forwards,
            alpha
        );

    Vector drift =
        terminalMeasureDrift(
            time,
            forwards,
            alpha
        );

    Vector effectiveRateShocks =
        effectiveForwardShocks(
            time,
            forwards,
            alpha,
            rateFactorShocks
        );

    nextForwards = forwards;
    nextAlpha = alpha;

    double sqrtDt =
        std::sqrt(
            dt
        );

    for (std::size_t i = 0;
         i < numberOfForwards_;
         ++i)
    {
        if (active[i] < 0.5)
        {
            continue;
        }

        double diffusion =
            dotProduct(
                loadings[i],
                rateFactorShocks
            )
            *
            sqrtDt;

        double proposedForward =
            forwards[i]
            +
            drift[i]
            *
            dt
            +
            diffusion;

        double minimumForward =
            -sabrParameters_.shift()[i]
            +
            1.0e-8;

        nextForwards[i] =
            std::max(
                proposedForward,
                minimumForward
            );

        double rho =
            sabrParameters_.rho()[i];

        double independentWeight =
            std::sqrt(
                std::max(
                    1.0 - rho * rho,
                    0.0
                )
            );

        double correlatedVolatilityShock =
            rho
            *
            effectiveRateShocks[i]
            +
            independentWeight
            *
            independentVolatilityShocks[i];

        double nu =
            sabrParameters_.nu()[i];

        double logIncrement =
            -0.5
            *
            nu
            *
            nu
            *
            dt
            +
            nu
            *
            sqrtDt
            *
            correlatedVolatilityShock;

        double proposedAlpha =
            alpha[i]
            *
            std::exp(
                logIncrement
            );

        proposedAlpha =
            std::max(
                proposedAlpha,
                sabrParameters_.alphaFloor()
            );

        proposedAlpha =
            std::min(
                proposedAlpha,
                sabrParameters_.alphaCap()
            );

        nextAlpha[i] =
            proposedAlpha;
    }

    validateState(
        nextForwards,
        nextAlpha
    );
}


Vector LMMSABRHybrid::discountFactorsFromForwards(
    const Vector& forwards,
    std::size_t startIndex
) const
{
    if (forwards.size() !=
        numberOfForwards_)
    {
        throw std::invalid_argument(
            "Forward vector has incorrect size."
        );
    }

    if (startIndex >
        numberOfForwards_)
    {
        throw std::out_of_range(
            "Invalid discount-factor start index."
        );
    }

    Vector discountFactors(
        numberOfForwards_ + 1,
        std::numeric_limits<double>::quiet_NaN()
    );

    discountFactors[startIndex] = 1.0;

    for (std::size_t i = startIndex;
         i < numberOfForwards_;
         ++i)
    {
        double denominator =
            1.0
            +
            accruals_[i]
            *
            forwards[i];

        if (denominator <= 0.0)
        {
            throw std::runtime_error(
                "Cannot reconstruct discount factors: "
                "1 + delta * L <= 0."
            );
        }

        discountFactors[i + 1] =
            discountFactors[i]
            /
            denominator;
    }

    return discountFactors;
}


double LMMSABRHybrid::terminalBondFromForwards(
    const Vector& forwards,
    std::size_t startIndex
) const
{
    Vector discountFactors =
        discountFactorsFromForwards(
            forwards,
            startIndex
        );

    return discountFactors.back();
}


double LMMSABRHybrid::initialTerminalBond() const
{
    return initialDiscountFactors_.back();
}


void LMMSABRHybrid::validateState(
    const Vector& forwards,
    const Vector& alpha
) const
{
    if (forwards.size() !=
        numberOfForwards_)
    {
        throw std::invalid_argument(
            "Forward state has incorrect size."
        );
    }

    if (alpha.size() !=
        numberOfForwards_)
    {
        throw std::invalid_argument(
            "Alpha state has incorrect size."
        );
    }

    for (std::size_t i = 0;
         i < numberOfForwards_;
         ++i)
    {
        if (!std::isfinite(forwards[i]) ||
            !std::isfinite(alpha[i]))
        {
            throw std::runtime_error(
                "The model state contains "
                "non-finite values."
            );
        }

        if (alpha[i] <= 0.0)
        {
            throw std::runtime_error(
                "Alpha must remain positive."
            );
        }

        if (forwards[i] +
            sabrParameters_.shift()[i]
            <= 0.0)
        {
            throw std::runtime_error(
                "Forward state violates "
                "L + shift > 0."
            );
        }

        if (1.0 +
            accruals_[i]
            *
            forwards[i]
            <= 0.0)
        {
            throw std::runtime_error(
                "Forward state violates "
                "1 + delta * L > 0."
            );
        }
    }
}


void LMMSABRHybrid::printSummary() const
{
    std::cout
        << "LMM-SABR Hybrid Model\n"
        << "------------------------------\n"
        << "Number of forwards: "
        << numberOfForwards_
        << "\n"
        << "Number of rate factors: "
        << numberOfFactors_
        << "\n"
        << "Volatility level: "
        << volatilityLevel_
        << "\n"
        << "Volatility decay: "
        << volatilityDecay_
        << "\n"
        << "Volatility floor: "
        << volatilityFloor_
        << "\n"
        << "Initial terminal bond: "
        << initialTerminalBond()
        << "\n";
}
